src/Control/Monad/Ghc.hs

```diff
@@ -60,7 +60,20 @@ instance (MonadIO m, MonadMask m) => MonadMask (GhcT m) where
         wrap g   = GhcT $ GHC.GhcT $ \s -> MTLAdapter (g s)
         unwrap m = unMTLA . GHC.unGhcT (unGhcT m)
 
-    uninterruptibleMask = mask
+    uninterruptibleMask f = wrap $ \s ->
+                              uninterruptibleMask $ \io_restore ->
+                                unwrap (f $ \m -> (wrap $ \s' -> io_restore (unwrap m s'))) s
+      where
+        wrap g   = GhcT $ GHC.GhcT $ \s -> MTLAdapter (g s)
+        unwrap m = unMTLA . GHC.unGhcT (unGhcT m)
```

Comment:
I also took the opportunity to fix the implementation of `uninterruptibleMask`. It was previously defined as `uninterruptibleMask = mask`, so calling `uninterruptibleMask body` was misleadingly running `body` in interruptible mode, not in uninterruptible mode. The difference is that some calls, like `threadDelay`, are intentionally marked as being interruptible even when running inside `mask`. As a result,

    > timeout 1000000 $ mask $ \_ -> threadDelay 3000000

terminates after 1 second because the timeout successfully interrupts the threadDelay after 1 second, while

    > timeout 1000000 $ uninterruptibleMask $ \_ -> lift $ threadDelay 3000000

terminates after 3 seconds, because the timeout has to wait until the `uninterruptibleMask` computation terminates before aborting the computation.

Without this fix,

    > timeout 1000000 $ runGhcT ... $ mask $ \_ -> lift $ threadDelay 3000000

and

    > timeout 1000000 $ runGhcT ... $ uninterruptibleMask $ \_ -> lift $ threadDelay 3000000

both terminate after 1 second, whereas with this fix, the second terminates after 3 seconds, as desired.